crates/nargo/src/ops/foreign_calls.rs

```diff
@@ -0,0 +1,139 @@
+    ReverseSequence,
+}
+
+impl std::fmt::Display for ForeignCall {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name())
+    }
+}
+
+impl ForeignCall {
+    pub(crate) fn name(&self) -> &'static str {
+        match self {
+            ForeignCall::Println => "println",
+            ForeignCall::PrintlnFormat => "println_format",
+            ForeignCall::Sequence => "get_number_sequence",
+            ForeignCall::ReverseSequence => "get_reverse_number_sequence",
+        }
+    }
+
+    pub(crate) fn lookup(op_name: &str) -> Option<ForeignCall> {
+        match op_name {
+            "println" => Some(ForeignCall::Println),
+            "println_format" => Some(ForeignCall::PrintlnFormat),
+            "get_number_sequence" => Some(ForeignCall::Sequence),
+            "get_reverse_number_sequence" => Some(ForeignCall::ReverseSequence),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn execute(
+        foreign_call: &ForeignCallWaitInfo,
+    ) -> Result<ForeignCallResult, ForeignCallError> {
+        let foreign_call_name = foreign_call.function.as_str();
+        match Self::lookup(foreign_call_name) {
+            Some(ForeignCall::Println) => {
+                Self::execute_println(&foreign_call.inputs)?;
+                Ok(foreign_call.inputs[0][0].into())
+            }
+            Some(ForeignCall::PrintlnFormat) => {
+                Self::execute_println_format(&foreign_call.inputs)?;
+                Ok(foreign_call.inputs[0][0].into())
+            }
+            Some(ForeignCall::Sequence) => {
+                let sequence_length: u128 = foreign_call.inputs[0][0].to_field().to_u128();
+
+                Ok(vecmap(0..sequence_length, Value::from).into())
+            }
+            Some(ForeignCall::ReverseSequence) => {
+                let sequence_length: u128 = foreign_call.inputs[0][0].to_field().to_u128();
+
+                Ok(vecmap((0..sequence_length).rev(), Value::from).into())
+            }
+            None => panic!("unexpected foreign call {:?}", foreign_call_name),
+        }
+    }
+
+    fn execute_println(foreign_call_inputs: &[Vec<Value>]) -> Result<(), ForeignCallError> {
+        // Fetch the abi type from the foreign call input
+        // The remaining input values should hold what is to be printed
+        let (abi_type, input_values) = fetch_abi_type(foreign_call_inputs)?;
+
+        // We must use a flat map here as each value in a struct will be in a separate input value
+        let mut input_values_as_fields =
+            input_values.iter().flat_map(|values| vecmap(values, |value| value.to_field()));
+        let decoded_value = decode_value(&mut input_values_as_fields, &abi_type)?;
+
+        let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
+        let output_string = serde_json::to_string_pretty(&json_value)
+            .map_err(|err| ForeignCallError::InputParserError(err.into()))?;
```

Comment:
Displaying these should be done with a `Display` or `Debug` implementation.

Reply:
done with a `Display` impl

Reply:
I was thinking that we'd have a `Display` implementation for `InputValue` as that's ultimately what we're wanting to display. I'd prefer to keep `JsonTypes` visible only within the `noirc_abi` crate.

Reply:
https://github.com/noir-lang/noir/issues/1988 I made an issue here for now and will submit a proposal. I think we will need a wrapper around `InputValue` to use `Display` here